# BioEng C112 (Molecular Biomechanics and Mechanobiology of the Cell) Course Materials
Scripts, Jupyter Notebooks, homeworks, and other material of molecular biomechanics course. 

## Notebooks

Notebook  | Description | 
---------------| ---------------
| [*Introduction To Python - 1*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/python/1%20-%20Basics.ipynb)| Basics of Python |
| [*Introduction To Python - 2*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/python/2%20-%20String%2C%20Lists%2C%20Loops.ipynb)| Strings, Lists, For Loops in Python |
| [*Introduction To Python - 3*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/python/3%20-%20Functions%2C%20Dictionaries%2C%20Args%2C%20Scopes.ipynb)| Functions, Dictionaries, Arguments, and Scopes |
| [*Introduction To Python - 4*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/python/4%20-%20Modules%2C%20Files%2C%20CSVs.ipynb)| Modules, File I/O (Specially CSV) in Python |
| [*Genomics with BioPython*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/genomics/genomics.ipynb)| Genomics Sequence Analysis with BioPython |
| [*Natural Language Processing*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/deep%20learning/Word2Vec.ipynb)| NLP basics with Gensim, Word2Vec Introduction |
| [*ProtVec*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/proteomics/ProtVec.ipynb)| Working with pre-trained protein vectors (ProtVec) |
| [*Molecular Dynamics*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/proteomics/MD.ipynb)| Molecular Dynamics introduction with MDAnalysis |
| [*Normal Mode Analysis*](https://nbviewer.org/github/Naghipourfar/molecular-biomechanics/blob/main/proteomics/NMA.ipynb)| Normal Mode Analysis with ProDy |




## References

* Cock, P.J.A. et al. [Biopython](https://biopython.org/): freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics 2009 Jun 1; 25(11) 1422-3 https://doi.org/10.1093/bioinformatics/btp163 pmid:19304878

* [Molecular Dynamics simulations in Python](https://klyshko.github.io/teaching/2019-03-01-teaching), Graduate course lecture, University of Toronto Missisauga, Department of Chemical and Physical Sciences, Taught by Eugene Klyshko, 2019

* [CS224n: Natural Language Processing with Deep Learning (Stanford)](https://web.stanford.edu/class/cs224n/)

* Asgari E, Mofrad MRK (2015) Continuous Distributed Representation of Biological Sequences for Deep Proteomics and Genomics. PLOS ONE 10(11): e0141287. https://doi.org/10.1371/journal.pone.0141287

* Jumper, J., Evans, R., Pritzel, A. et al. Highly accurate protein structure prediction with AlphaFold. Nature 596, 583–589 (2021). https://doi.org/10.1038/s41586-021-03819-2

* Ahmet Bakan, Lidio M. Meireles, Ivet Bahar, [ProDy](http://prody.csb.pitt.edu/): Protein Dynamics Inferred from Theory and Experiments, Bioinformatics, Volume 27, Issue 11, 1 June 2011, Pages 1575–1577, https://doi.org/10.1093/bioinformatics/btr168

* Python programming for biologists, Tel-Aviv university, 0411-3122, Yoav Ram, Lior Glick, Spring 2015

* Rehurek, R., & Sojka, P. (2011). [Gensim–python](https://radimrehurek.com/gensim/) framework for vector space modelling. NLP Centre, Faculty of Informatics, Masaryk University, Brno, Czech Republic, 3(2).

* Wolf, F., Angerer, P. & Theis, F. [SCANPY](https://scanpy.readthedocs.io/en/stable/): large-scale single-cell gene expression data analysis. Genome Biol 19, 15 (2018). https://doi.org/10.1186/s13059-017-1382-0

* Harris, C.R., Millman, K.J., van der Walt, S.J. et al. Array programming with [NumPy](https://numpy.org/). Nature 585, 357–362 (2020). DOI: 10.1038/s41586-020-2649-2. (Publisher link).

* Jeff Reback, Wes McKinney, jbrockmendel, Joris Van den Bossche, Tom Augspurger, Phillip Cloud, gfyoung, Sinhrks, Adam Klein, Matthew Roeschke, Simon Hawkins, Jeff Tratner, Chang She, William Ayd, Terji Petersen, Marc Garcia, Jeremy Schendel, Andy Hayden, MomIsBestFriend, … Mortada Mehyar. (2020). pandas-dev/pandas: [Pandas](https://pandas.pydata.org/) 1.0.3 (v1.0.3). Zenodo. https://doi.org/10.5281/zenodo.3715232

* J. D. Hunter, "[Matplotlib](https://matplotlib.org/): A 2D Graphics Environment", Computing in Science & Engineering, vol. 9, no. 3, pp. 90-95, 2007

* Fernando Pérez, Brian E. Granger, [IPython](https://ipython.org): A System for Interactive Scientific Computing, Computing in Science and Engineering, vol. 9, no. 3, pp. 21-29, May/June 2007, doi:10.1109/MCSE.2007.53. URL: https://ipython.org

* R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler, D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein. [MDAnalysis](https://www.mdanalysis.org/): A Python package for the rapid analysis of molecular dynamics simulations. In S. Benthall and S. Rostrup, editors, Proceedings of the 15th Python in Science Conference, pages 98-105, Austin, TX, 2016. SciPy, doi:10.25080/majora-629e541a-00e.

* N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein. MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations. J. Comput. Chem. 32 (2011), 2319-2327, doi:10.1002/jcc.21787. PMCID:PMC3144279







